### ISSUE TITLE
Problems with tf.keras.metrics.CategoricalAccuracy?

### ISSUE THREAD

**Issue by shraddhazpy** [tensorflow/tensorflow#46953]
Hi

As per the documentation in https://www.tensorflow.org/api_docs/python/tf/keras/metrics/CategoricalAccuracy , I am to understand that this metric is expecting y_true to be one hot encoded. Now, taking an example from the docs:

```
m = tf.keras.metrics.CategoricalAccuracy()
m.update_state([[0, 0, 1], [0, 1, 0]], [[0.1, 0.9, 0.8],[0.05, 0.95, 0]])
m.result().numpy() 
```
This gives me the desired result.
But when I DO NOT one hot encode the y_true data and just use the labels(something for which SparseCategoricalAccuracy is there).
Code shown below:
```
m = tf.keras.metrics.CategoricalAccuracy()
m.update_state([[2], [1]], [[0.1, 0.9, 0.8],[0.05, 0.95, 0]]) 
m.result().numpy()
```
Instead of raising an exception, it gives me a result.
Was this designed intentionally?
If no, can we do something about it cause it'd be really helpful to intimate the user about a possible miss(not one hot encoding the y_true labels) on their part.

**Comment by AdityaKane2001** [comment]
I am not sure whether this answer is correct, but please hear me out.

1. If you go to [source](https://github.com/tensorflow/tensorflow/blob/85c8b2a817f95a3e979ecd1ed95bff1dc1335cff/tensorflow/python/keras/metrics.py#L733) of CategoricalAccuracy, you will see that it is just an inherited class which has `categorical_accuracy()` function defined on line 3250 ([here](https://github.com/tensorflow/tensorflow/blob/85c8b2a817f95a3e979ecd1ed95bff1dc1335cff/tensorflow/python/keras/metrics.py#L3250)). That is a straightforward peice of code, which will work exactly as expected. 
2.  In the ` MeanMetricWrapper` class (line [572](https://github.com/tensorflow/tensorflow/blob/85c8b2a817f95a3e979ecd1ed95bff1dc1335cff/tensorflow/python/keras/metrics.py#L572)),  you will see that in `update_state()` :
 ` y_pred, y_true = losses_utils.squeeze_or_expand_dimensions(
        y_pred, y_true)`
3. Now if you go to [`squeeze_or_expand_dimensions()` ](https://github.com/tensorflow/tensorflow/blob/85c8b2a817f95a3e979ecd1ed95bff1dc1335cff/tensorflow/python/keras/utils/losses_utils.py#L99), you will see that the function checks if the rank of both tensors differs by exactly 1, and if yes, it expands or squeezes the tensors. Due to this, `[[2],[1]]` becomes a valid input, and thus CategoricalAccuracy returns 0. Please find gist [here](https://colab.research.google.com/gist/AdityaKane2001/9bce96510c0b6e5e237af1b80fcb13e9/46953.ipynb).

Hope this helps.

**Comment by shraddhazpy** [comment]
@AdityaKane2001 Thanks for your comment. I saw the source code and I agree with you as to why it is the way it is. What I want to know is , is this the way it is supposed be **functionally**.

**Comment by AdityaKane2001** [comment]
Following is the categorical accuracy function:

```
def categorical_accuracy(y_true, y_pred):
  """Calculates how often predictions matches one-hot labels.
  Standalone usage:
              ...
  Args:
              ...
  Returns:
             ...
  """
  return math_ops.cast(
      math_ops.equal(
          math_ops.argmax(y_true, axis=-1), math_ops.argmax(y_pred, axis=-1)),
      K.floatx())

```

I think the way it works is not intentional, and the function should ideally check the y_pred Tensor to be of one hot encoded before calculating the accuracy.

**Comment by AdityaKane2001** [comment]
Requesting to close this issue, as it has been solved in [#48000 ](https://github.com/tensorflow/tensorflow/pull/48000)

**Comment by chenmoneygithub** [comment]
Thanks!
